### PR TITLE
Design doc: Importing our prometheus metrics into system/log tables

### PR DESCRIPTION
**Pull request by antifuchs**
[rendered](https://github.com/antifuchs/materialize/blob/asf/rfc--prometheus-source/doc/developer/design/20210311_prometheus_source.md). 

I took the writeup from https://github.com/MaterializeInc/materialize/issues/1696#issuecomment-796898988 and turned it into a design doc. This doc fleshes out the design for a source that reads from prometheus endpoints and makes them into materialized sources.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6101)

### CODE REVIEW

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [inline doc/developer/design/20210311_prometheus_source.md:78]: I view "figuring out how to aggregate the metric properly" as a good thing. Perhaps if there are easy helpers that "just do the right thing", we can make it easy to add new metrics *and* have them aggregated easily?

 I have this strange world view that all metrics should presented as aggregations, particularly when they are intended for human consumption. There's only so much data that can be presented on a display / read by a human and, for most metrics use cases, that means summarizing events. So, that's my verbose way of saying that I think we should require all metrics have proper aggregation semantics.

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [inline doc/developer/design/20210311_prometheus_source.md:3]: This is about implementing a generic Prometheus metrics scraper in Materialize, correct? This is not about exposing internal system tables as Prometheus metrics to be scraped, correct?

Perhaps we should call this `Prometheus metrics source / scraper`?

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [inline doc/developer/design/20210311_prometheus_source.md:70]: This seems to to be about exposing internal Prometheus metrics directly into a system table. That's a different feature, right?

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [inline doc/developer/design/20210311_prometheus_source.md:135]: I think the generic scraper would be awesome, particularly for being able to write quality alerts from Prometheus metrics that need to join data against external sources.

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [inline doc/developer/design/20210311_prometheus_source.md:138]: It seems to me that we should write the data in normalized tables, particularly since this lends itself better to having an event-sourced writer pattern for updating metadata and having the updates reflected downstream.

**Review by cirego — COMMENTED** [review]
This is awesome! I'd like to understand better if the use case is for ingesting generic Prometheus metrics into Materialize or if the use case is to expose materialized Prometheus metrics via system tables.

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:3]: Yep, that's correct. Renaming sounds good (I thought it was pretty much implicit in the nature of the source, but good to make explicit!)

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:78]: Ah, I cut off some of the reasoning here, maybe need to expand it: Much like the main difficulty with implementing #5825, the way we handle metrics in code right now makes it very hard (possibly impossible with the current stream abstractions) to pass a channel like that down to, say, the code responsible for serializing to the pg wire protocol. Maybe not insurmountable! But definitely a very heavy lift.

The other thing is that in my experience, it's preferable to provide one "reasonable-enough" metrics-providing abstraction, over making them choose between / hook together several options. This design would let us keep providing that reasonable-enough interface via prometheus internally (:

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:70]: What I want to do here that's the one non-negotiable requirement is getting our own prometheus metrics into system tables. That can happen with or without a prometheus scraper, so not really a different feature. Same end goal, different implementation mechanism.

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [reply doc/developer/design/20210311_prometheus_source.md:78]: I totally agree with having there be a single abstraction! My main concern here is that our chosen abstraction considers aggregations from the start.

**Comment by benesch** [comment]
Thanks very much for writing all of this up, @antifuchs! I'm sorry I haven't been paying more attention to this discussion (perils of there being so many teams these days), and apologies in advance if what I'm about to write is comes across as overly negative—I am in fact very glad that someone is finally thinking about this!

I share Chris's question about whether the primary goal here is to support a Prometheus source for use with external Prometheus instances, or whether the primary goal is to support exposing our Prometheus metrics in a system table. 

Adding support for Prometheus sources seems broadly useful, but not in line with the technical vision we've got laid out for the next few months. I'd really like to avoid adding new externally-usable source types for a bit so we can focus on shoring up the source types that we already have.

I quite like the alternative you proposed called "Do not scrape arbitrary prometheus endpoints, instead scrape our internal metrics registry". Is it really blocked on #5825? Can't we just call `gather_metrics` from within the dataflow layer?

Separately, we just need to fix #5825. I have been beating this drum for months to no avail, and the problem just keeps getting worse. I'll work with @umanwizard to figure out how to get this work scheduled, because I don't like the idea of this technical debt continuing to accrete. Doubly so if it impacts the solution here.

> Besides that, I think it's somewhat nice to represent this data as a source that we poll, and internally, our metrics could still be realized like this once a fix for #5825 lands.

I think this makes sense, but I think we can keep it as an implementation detail? Like, internally, we can have a "prometheus source", but not expose it to users. And to start it will only need to call `gather_metrics` rather than making an HTTP request, which will reduce the surface area of possible failures.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:11]: I don't believe that this is a current goal -- and your comment here https://github.com/MaterializeInc/materialize/issues/1696#issuecomment-796949699 suggests that originally you felt the same way?

I think that this in particular adds a lot of complexity that we should strive to avoid for v0.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [reply doc/developer/design/20210311_prometheus_source.md:3]: I think I misunderstood our conversation, I thought that we were discussing just scraping our internal metrics (via registry::gather, aka your second alternative), without hitting any external endpoints. That seems like it would satisfy the requirements of getting _our_ metrics data into system tables without adding a lot of other complexity.

New prometheus source types are, I think, currently just skunkworks projects (which I would be interested in chatting about!) or deprioritized for a quarter or two.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:19]: Strong agree on this one for this design, I think that's an entirely separate project.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:28]: Sorry, I think we got wires crossed, I thought that we agreed that your alternative "Do not scrape arbitrary prometheus endpoints, instead scrape our internal metrics registry" was what we were discussing. I think that that is the better alternative for this phase of the project.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:84]: This isn't correct: we can 100% get all metrics data into our system tables without that issue, if anything doing it before we finish #5825 will be microscopically easier because we won't need to pass the registry into 1 more place.

**Comment by quodlibetor** [comment]
@benesch 

> Is it really blocked on #5825? Can't we just call gather_metrics from within the dataflow layer?

You're correct, we can just call gather_metrics, this is unrelated to whether we have a static registry or thread one through the code base.

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:84]: @benesch has voiced the same point, and thinking about this more - I agree! Now that I think about this, the causality in the paragraph above is the wrong way around: scraping of the metrics registry would make it microscopically harder to fix #5825, much like adding any single metric does (-:

So, given these points (by you & Nikhil), I feel that we should not make an externally-scraping source yet but instead go with this design: Feeding a dataflow from the global (or, later, the "materialized") registry and skipping the "poll a random prometheus endpoint" bits.

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:11]: I've been wavering between the two: Retaining some history would make it easier to aggregate things, but on the other hand, make it a bit harder to query for the "current" value.

I feel it's probably better to have some (minor) access to historic values (so users can use query tools to chart the progression of a metric) than to throw that info away, and given that we're making timely dataflows from the information, it felt like this information would be easy to retain & process with (especially since we need a timestamp on a source anyway).

I'm probably missing the source of complexity that you're seeing - what do you think would make this difficult?

**Review by antifuchs — COMMENTED** [review]
OK, that's a fantastic set of feedback, thanks everyone!

I hear you on the point that scraping an external prometheus endpoint is not a thing we want to do now (and surveying the [landscape of libraries/library](https://crates.io/crates/prometheus-parse) we have available to do so, it seems the community agrees (-:

I'll rework this design doc to scope to the [second alternative, scraping our internal metrics registry](https://github.com/antifuchs/materialize/blob/e8c2da1bf66bbb7f3e87d0193121817eb6b93d1b/doc/developer/design/20210311_prometheus_source.md#do-not-scrape-arbitrary-prometheus-endpoints-instead-scrape-our-internal-metrics-registry).

Some questions about the data model remain, but I think we can discuss this once the major boulder is out of the way!

**Comment by antifuchs** [comment]
Updated to bring the proposed implementation in line with what you all proposed (outdating a bunch of comments in the process, I suspect).

The things that I'm aware of that need hashing out are:
- [ ] should we retain historical values for any amount of time? I think we can ~easily model the removal of old values with temporal filters (thanks @ruchirK for the primer on timestamps!).
- [ ] are we aligned on the normalized form of these tables? (I got one yes vote on normalized, and am happy to keep them normalized)

File: doc/developer/design/20210311_prometheus_source.md
Comment (cirego) [reply doc/developer/design/20210311_prometheus_source.md:11]: Perhaps we should make this goal optional or a stretch goal? I agree that it adds complexity and it would be good to know if we consider "values over time" a hard requirement.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:38]: I'm not sure that `timestamp` makes sense as something to include unless we're going to be precise  about when items were actually inserted/removed from the tables.

**Review by cirego — APPROVED** [review]
I vote that, if it's easy to model values over time, we go for it!

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:50]: I think that stating that we will expose this configuration parameter to end users should be left for some follow up work.

That said, if you're going to define it here, what exactly is the API? Is it a CLI arg or a SQL function or a system variable that can be SET? I think a CLI arg is probably the most straightforward thing to implement, and matches things like rdkafka log frequency.

Comment (quodlibetor) [inline doc/developer/design/20210311_prometheus_source.md:50]: I think a default poll interval of 1 or 5 seconds is more likely to be pleasant hopefully without meaningfully impacting performance of the system as a whole.

**Review by quodlibetor — COMMENTED** [review]
> should we retain historical values for any amount of time?

IMO this shouldn't be part of the initial PR, it should have its own separate design discussion and shouldn't block usage of the feature. In particular, it's unclear how timestamps should be assigned, how they'd apply to data in other system tables, and whether we should insert new records if the only thing that has changed is the timestamp.

> are we aligned on the normalized form of these tables?

I think that the normalized form looks good, modulo my comment about the `timestamp` field being unnecessary and possibly misleading.

**Comment by antifuchs** [comment]
> In particular, it's unclear how timestamps should be assigned, how they'd apply to data in other system tables, and whether we should insert new records if the only thing that has changed is the timestamp.

All these are good call-outs. I think the straight-forward ways to answer that are, in order:
* We use the real-time stamp of when the task scraped the registry. That's also the ts we'd use on the dataflow triple, so should adapt well to temporal filters, too.
* Since the prometheus metrics are pretty much unrelated to the data we have on system tables, there's not really any way to join them anyway. In discussing this with @ruchirK earlier, he proposed to make a view containing only the most recent data (thinned out with a lateral join or whatever low-level operator makes this easiest). Might be worth it.
* Yes, definitely insert unchanged values. There is a qualitative difference between "no data point" and "data point is present and the same as the previous one", and we should only show absence of a reading if the reading was absent.

I am happy to strike this out from the design if you're opposed to their inclusion, but these items above were what I was going by in the design & they seemed somewhat clear to me. (Also happy to make them explicit in the doc!)

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:50]: I was going with the [default prometheus poll interval](https://prometheus.io/docs/prometheus/latest/configuration/configuration/), 60s. Do we configure our metric pipeline for shorter intervals? Then this definitely ought to change.

Comment (quodlibetor) [reply doc/developer/design/20210311_prometheus_source.md:50]: Yeah our cloud prometheus polls typically around every 15 seconds,b ut I don't know that that's necessarily the right metaphor for what this data is meant to reflect. The cloud prometheus polling interval needs to respect the fact that a single prometheus node is often responsible for scraping tens to hundreds or thousands of instances, and the data is meant to be pretty coarse.

This internal item is meant to accurately reflect what is going on internally in our system, and should generally be able to scrape in a couple milliseconds. It should generally actually probably match our [`--introspection-frequency`](https://github.com/MaterializeInc/materialize/blob/44f1e7168db258f20565363546ef5646dbaf1ba3/src/materialized/src/bin/materialized/main.rs#L94-L102) cli arg. That argument used to be named --logging-granularity or something, but the new name seems like it should affect this. (cc @benesch for his opinion)

**Comment by quodlibetor** [comment]
I think that your position on the times makes sense, they match the way that prometheus tracks value times. I think I was hoping one day to have finer-grained data capable of being expressed -- therefore including intermediate timestamps between polls -- but having a maximum coarseness of the poll interval is reasonable.

I still think that the first version of this PR work will be easier to implement if it doesn't include any historical data, but I could be wrong about that.

**Comment by quodlibetor** [comment]
Thinking about the "only current" vs "include historical" question, I suppose that not including historical data means that we need to keep track of everything that was inserted in the previous poll so that we can negate it on the subsequent poll, whereas using the filter out old data logic reduces the need to handle that kind of thing explicitly.

File: doc/developer/design/20210311_prometheus_source.md
Comment (cuongdo) [reply doc/developer/design/20210311_prometheus_source.md:50]: The open-source databases I've used update their metrics often, so +1 to updating more often. However, does INSERTing into Materialize often cause performance issues? I vaguely recall some performance limitations for tables.

File: doc/developer/design/20210311_prometheus_source.md
Comment (umanwizard) [inline doc/developer/design/20210311_prometheus_source.md:66]: We already have the timely logs which sound very similar to what you're describing here -- what is the difference between this approach and that?

**Comment by antifuchs** [comment]
> not including historical data means that we need to keep track of everything that was inserted in the previous poll so that we can negate it on the subsequent poll, whereas using the filter out old data logic reduces the need to handle that kind of thing explicitly.

Yep - canceling away the old values in the "no historical readings" case was what I am worried about as well - maybe there is a super clever way to do it, but I can't come up one that doesn't require us to keep the whole last batch of readings around... which doesn't sound too terrible, but if we can do it & it's easier to do that way (which it seems it is!), I figure we should provide historical values too (:

File: doc/developer/design/20210311_prometheus_source.md
Comment (antifuchs) [reply doc/developer/design/20210311_prometheus_source.md:66]: The main (only real) difference is the interface at the point of metric value emission - currently, things using the prometheus interface have an easier time, e.g. tracking histograms (which we have to do [faiiirly manually in timely logging code](https://github.com/MaterializeInc/materialize/blob/ebed079124da5fdcc50a0e97be947b4c39fac3c6/src/dataflow/src/logging/timely.rs#L331-L338)). That, together with the constraints mentioned in the tracking issue (https://github.com/MaterializeInc/materialize/issues/1696#issuecomment-587283328), my understanding is we will want to retain prometheus as an external and internal interface (external to ensure that materialized run can be measured right by our users, internal to ensure the metrics we track actually make it to the external interface).

So I think we'd have two options for getting the same values into prometheus's external interface & our internal log tables: 
* importing the metric values from the prometheus registry via some public interface (the design proposed here), or
* doing the work twice at the point of metric value emission: incrementing a prometheus counter & sending an event down a channel that gets counted / processed into a log table entry (this alternative approach).

I don't much love the second option because it means we effectively do the work twice and we have to ~copy the behavior of prometheus or risk questions about which number is right (the old adage about a man with two wristwatches never knowing what time it is).

**Comment by antifuchs** [comment]
I feel this may have stalled out a little bit. Are there any blockers/open questions that people have that still need figuring out? Otherwise, I'd default to merging & getting started on implementing this (-:

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [reply doc/developer/design/20210311_prometheus_source.md:50]: > does INSERTing into Materialize often cause performance issues

I don't expect this design to have those issues. We had some write-amplification issues when we included some of our current prometheus metrics as table metrics. Doing that meant that parsing 1 row of data from an external source could cause ~10 writes to our system tables for various counters, and it would be worse now because we've added more metrics. With the design in this doc we will increase the number of writes based on the number of metrics, but always at `--introspection-frequency` intervals, so basically we'll be inserting compacted metrics not row-by-row metrics.

File: doc/developer/design/20210311_prometheus_source.md
Comment (quodlibetor) [reply doc/developer/design/20210311_prometheus_source.md:66]: This is also a reference to the design in #5277, which had the goal of making it easier to insert into timely logs from places in out codebase that aren't the `dataflow` crate.

**Review by quodlibetor — APPROVED** [review]

**Comment by quodlibetor** [comment]
I think this is good to go modulo the small comments from slack!

One thing to keep in mind is that I think that if the historical data implementation becomes complex it's fine to punt on it for the first PR.

**Comment by antifuchs** [comment]
Updated, and noted that we may punt the retention bit if it's too complex (:

**Review by cirego — APPROVED** [review]
This looks great to me!

**Pull request merged**
